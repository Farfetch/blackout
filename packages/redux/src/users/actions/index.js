/**
 * Users actions.
 *
 * @module users/actions
 * @category Users
 * @subcategory Actions
 */

export { default as createContact } from './createContact';
export { default as createGuestUser } from './createGuestUser';
export { default as fetchBenefits } from './fetchBenefits';
export { default as fetchContact } from './fetchContact';
export { default as fetchContacts } from './fetchContacts';
export { default as fetchCredit } from './fetchCredit';
export { default as fetchCreditMovements } from './fetchCreditMovements';
export { default as fetchGuestUser } from './fetchGuestUser';
export { default as fetchPreferences } from './fetchPreferences';
export { default as fetchTitles } from './fetchTitles';
export { default as fetchUser } from './fetchUser';
export { default as removeContact } from './removeContact';
export { default as updateContact } from './updateContact';
export { default as updatePreferences } from './updatePreferences';
export { default as updateUser } from './updateUser';
