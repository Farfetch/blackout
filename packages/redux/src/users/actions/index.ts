/**
 * Users actions.
 */

export { default as createContact } from './createContact';
export { default as createGuestUser } from './createGuestUser';
export { default as fetchBenefits } from './fetchBenefits';
export { default as fetchContact } from './fetchContact';
export { default as fetchContacts } from './fetchContacts';
export { default as fetchCredit } from './fetchCredit';
export { default as fetchCreditMovements } from './fetchCreditMovements';
export { default as fetchGuestUser } from './fetchGuestUser';
export { default as fetchPreferences } from './fetchPreferences';
export { default as fetchTitles } from './fetchTitles';
export { default as fetchUser } from './fetchUser';
export { default as removeContact } from './removeContact';
export { default as updateContact } from './updateContact';
export { default as setPreferences } from './setPreferences';
export { default as setUser } from './setUser';
export { default as fetchUserAttributes } from './fetchUserAttributes';
export { default as createUserAttributes } from './createUserAttributes';
export { default as fetchUserAttribute } from './fetchUserAttribute';
export { default as setUserAttribute } from './setUserAttribute';
export { default as updateUserAttribute } from './updateUserAttribute';
export { default as removeUserAttribute } from './removeUserAttribute';
export { default as createPhoneTokens } from './createPhoneTokens';
export { default as createPhoneTokenValidations } from './createPhoneTokenValidations';
export { default as createPhoneNumberValidations } from './createPhoneNumberValidations';
export { default as fetchPersonalIds } from './fetchPersonalIds';
export { default as createPersonalIds } from './createPersonalIds';
export { default as fetchDefaultPersonalId } from './fetchDefaultPersonalId';
export { default as setDefaultPersonalId } from './setDefaultPersonalId';
export { default as createPersonalIdImage } from './createPersonalIdImage';
export { default as removePersonalId } from './removePersonalId';
export { default as fetchPersonalId } from './fetchPersonalId';
export { default as updatePersonalId } from './updatePersonalId';
export { default as resetUser } from './resetUser';
export { default as resetUserState } from './resetUserState';
