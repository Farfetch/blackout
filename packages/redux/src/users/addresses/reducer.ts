import * as actionTypes from './actionTypes';
import { AnyAction, combineReducers, Reducer } from 'redux';
import omit from 'lodash/omit';
import produce from 'immer';
import reducerFactory, {
  createReducerWithResult,
} from '../../helpers/reducerFactory';
import type {
  AddressEntity,
  AddressesEntity,
} from '../../entities/types/addresses.types';
import type {
  FetchUserAddressesSuccessAction,
  FetchUserAddressSuccessAction,
  UserAddressesState,
} from './types';
import type { StoreState } from '../../types';

export const INITIAL_STATE: UserAddressesState = {
  error: null,
  isLoading: false,
  result: null,
  addresses: {
    error: null,
    isLoading: false,
  },
  address: {
    error: {},
    isLoading: {},
  },
  /* Used for operations related with the default address that
    have a result associated, such as getting the default contact address */
  defaultAddressDetails: {
    error: null,
    isLoading: false,
    result: null,
  },
};

// Goes through the addresses list and returns the default `shipping|billing|contact` address details. The
// address returned is the one passed in `prop` param, i.e. isCurrentShipping, isCurrentBilling or isCurrentPreferred.
// If no default address with that prop exists, null is returned.
export const getDefaultAddress = (
  addressesList: AddressesEntity,
  prop: 'isCurrentShipping' | 'isCurrentBilling' | 'isCurrentPreferred',
): AddressEntity | null | undefined => {
  for (const key in addressesList) {
    const address = addressesList[key];

    if (address && address[prop]) {
      return address;
    }
  }
  return null;
};

const result = (
  state = INITIAL_STATE.result,
  action: AnyAction,
): UserAddressesState['result'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ADDRESS_SUCCESS: {
      const fetchedAddressId = (action as FetchUserAddressSuccessAction).payload
        .result;

      if (!state) {
        return [fetchedAddressId];
      }

      if (!state.includes(fetchedAddressId)) {
        return [...state, fetchedAddressId];
      }

      return state;
    }
    case actionTypes.FETCH_USER_ADDRESSES_SUCCESS:
      return (action as FetchUserAddressesSuccessAction).payload.result;
    case actionTypes.CREATE_USER_ADDRESS_SUCCESS:
      if (!state) {
        return [action.meta.addressId];
      }

      return [...state, action.meta.addressId];
    case actionTypes.REMOVE_USER_ADDRESS_SUCCESS: {
      return (
        state && state.filter(addressId => addressId !== action.meta.addressId)
      );
    }
    default:
      return state;
  }
};

const error = (
  state = INITIAL_STATE.error,
  action: AnyAction,
): UserAddressesState['error'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_ADDRESSES_FAILURE:
    case actionTypes.CREATE_USER_ADDRESS_FAILURE:
    case actionTypes.UPDATE_USER_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_USER_ADDRESS_REQUEST:
    case actionTypes.FETCH_USER_ADDRESSES_REQUEST:
    case actionTypes.CREATE_USER_ADDRESS_REQUEST:
    case actionTypes.UPDATE_USER_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: AnyAction,
): UserAddressesState['isLoading'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ADDRESSES_REQUEST:
    case actionTypes.FETCH_USER_ADDRESS_REQUEST:
    case actionTypes.CREATE_USER_ADDRESS_REQUEST:
    case actionTypes.UPDATE_USER_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return true;
    case actionTypes.FETCH_USER_ADDRESSES_FAILURE:
    case actionTypes.FETCH_USER_ADDRESSES_SUCCESS:
    case actionTypes.FETCH_USER_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_ADDRESS_SUCCESS:
    case actionTypes.CREATE_USER_ADDRESS_FAILURE:
    case actionTypes.CREATE_USER_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_USER_ADDRESS_FAILURE:
    case actionTypes.UPDATE_USER_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.CREATE_USER_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const id = action.payload.result;
    const createdAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState) {
        draftState = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...createdAddress },
      };
    });
  },
  [actionTypes.UPDATE_USER_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const id = action.payload.result;
    const updatedAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState) {
        draftState = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...updatedAddress },
      };
    });
  },
  [actionTypes.REMOVE_USER_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;
    const currentAddresses = state?.addresses;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      draftState.addresses = omit(currentAddresses, addressId);

      return draftState;
    });
  },
  [actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;
    // Get prev default address so it can later be unmarked as the default
    const prevCurrentShippingAddress = getDefaultAddress(
      state?.addresses,
      'isCurrentShipping',
    );

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      if (prevCurrentShippingAddress) {
        // Unmark previous shipping address as default
        const prevCurrentShippingAddressStore =
          addresses[prevCurrentShippingAddress.id];
        if (prevCurrentShippingAddressStore) {
          prevCurrentShippingAddressStore.isCurrentShipping = false;
        }
      }

      // Select the selected address as default
      const newDefaultShippingAddress = addresses[addressId];
      if (newDefaultShippingAddress) {
        newDefaultShippingAddress.isCurrentShipping = true;
      }

      return draftState;
    });
  },
  [actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;
    // Get prev default address so it can later be unmarked as the default
    const prevCurrentBillingAddress = getDefaultAddress(
      state?.addresses,
      'isCurrentBilling',
    );

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      if (prevCurrentBillingAddress) {
        // Unmark previous billing address as default
        const prevCurrentBillingAddressStore =
          addresses[prevCurrentBillingAddress.id];
        if (prevCurrentBillingAddressStore) {
          prevCurrentBillingAddressStore.isCurrentBilling = false;
        }
      }

      // Select the selected address as default
      const newDefaultBillingAddress = addresses[addressId];
      if (newDefaultBillingAddress) {
        newDefaultBillingAddress.isCurrentBilling = true;
      }

      return draftState;
    });
  },
  [actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevCurrentContactAddress = getDefaultAddress(
      state?.addresses,
      'isCurrentPreferred',
    );

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      if (prevCurrentContactAddress) {
        // Unmark previous contact address as default
        const prevCurrentContactAddressStore =
          addresses[prevCurrentContactAddress.id];

        if (prevCurrentContactAddressStore) {
          prevCurrentContactAddressStore.isCurrentPreferred = false;
        }
      }

      // Select the selected address as default
      const newDefaultContactAddress = addresses[addressId];

      if (newDefaultContactAddress) {
        newDefaultContactAddress.isCurrentPreferred = true;
      }

      return draftState;
    });
  },
  [actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      // Unmark the selected address as default
      const defaultContactAddress = addresses[addressId];

      if (defaultContactAddress) {
        defaultContactAddress.isCurrentPreferred = false;
      }

      return draftState;
    });
  },
};

export const address = (
  state = INITIAL_STATE.address,
  action: AnyAction,
): UserAddressesState['address'] => {
  switch (action.type) {
    case actionTypes.CREATE_USER_ADDRESS_REQUEST:
      return {
        isLoading: { ...state.isLoading },
        error: {},
      };
    case actionTypes.FETCH_USER_ADDRESS_REQUEST:
    case actionTypes.UPDATE_USER_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: true,
        },
        error: {},
      };
    case actionTypes.FETCH_USER_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_USER_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_USER_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
      };
    case actionTypes.FETCH_USER_ADDRESS_FAILURE:
    case actionTypes.UPDATE_USER_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
        error: {
          ...state.error,
          [action.meta.addressId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const addresses = reducerFactory(
  'FETCH_USER_ADDRESSES',
  INITIAL_STATE.addresses,
  actionTypes,
);

export const defaultAddressDetails = createReducerWithResult(
  'FETCH_USER_DEFAULT_CONTACT_ADDRESS',
  INITIAL_STATE.defaultAddressDetails,
  actionTypes,
);

export const getError = (
  state: UserAddressesState,
): UserAddressesState['error'] => state.error;
export const getIsLoading = (
  state: UserAddressesState,
): UserAddressesState['isLoading'] => state.isLoading;
export const getResult = (
  state: UserAddressesState,
): UserAddressesState['result'] => state.result;
export const getAddresses = (
  state: UserAddressesState,
): UserAddressesState['addresses'] => state.addresses;
export const getAddress = (
  state: UserAddressesState,
): UserAddressesState['address'] => state.address;
export const getDefaultAddressDetails = (
  state: UserAddressesState,
): UserAddressesState['defaultAddressDetails'] => state.defaultAddressDetails;

const reducer = combineReducers({
  error,
  isLoading,
  result,
  addresses,
  address,
  defaultAddressDetails,
});

/**
 * Reducer for addresses state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const addressesReducer: Reducer<UserAddressesState> = (state, action) => {
  if (action.type === actionTypes.RESET_USER_ADDRESSES) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};

export default addressesReducer;
