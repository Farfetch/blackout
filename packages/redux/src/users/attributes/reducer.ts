import * as actionTypes from './actionTypes';
import { type AnyAction, combineReducers, type Reducer } from 'redux';
import type * as T from './types';

export const INITIAL_STATE: T.UserAttributesState = {
  result: null,
  error: null,
  isLoading: false,
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: AnyAction,
): T.UserAttributesState['isLoading'] => {
  switch (action.type) {
    case actionTypes.CREATE_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTE_REQUEST:
    case actionTypes.REMOVE_USER_ATTRIBUTE_REQUEST:
    case actionTypes.SET_USER_ATTRIBUTE_REQUEST:
    case actionTypes.UPDATE_USER_ATTRIBUTE_REQUEST:
      return true;
    case actionTypes.CREATE_USER_ATTRIBUTES_FAILURE:
    case actionTypes.CREATE_USER_ATTRIBUTES_SUCCESS:
    case actionTypes.FETCH_USER_ATTRIBUTES_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTES_SUCCESS:
    case actionTypes.FETCH_USER_ATTRIBUTE_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTE_SUCCESS:
    case actionTypes.REMOVE_USER_ATTRIBUTE_FAILURE:
    case actionTypes.REMOVE_USER_ATTRIBUTE_SUCCESS:
    case actionTypes.SET_USER_ATTRIBUTE_FAILURE:
    case actionTypes.SET_USER_ATTRIBUTE_SUCCESS:
    case actionTypes.UPDATE_USER_ATTRIBUTE_FAILURE:
    case actionTypes.UPDATE_USER_ATTRIBUTE_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

type ErrorAction =
  | T.CreateUserAttributesFailureAction
  | T.FetchUserAttributeFailureAction
  | T.FetchUserAttributesFailureAction
  | T.RemoveUserAttributeFailureAction
  | T.SetUserAttributeFailureAction
  | T.UpdateUserAttributeFailureAction;

const error = (
  state = INITIAL_STATE.error,
  action: AnyAction,
): T.UserAttributesState['error'] => {
  switch (action.type) {
    case actionTypes.CREATE_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTE_REQUEST:
    case actionTypes.REMOVE_USER_ATTRIBUTE_REQUEST:
    case actionTypes.SET_USER_ATTRIBUTE_REQUEST:
    case actionTypes.UPDATE_USER_ATTRIBUTE_REQUEST:
      return INITIAL_STATE.error;
    case actionTypes.CREATE_USER_ATTRIBUTES_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTES_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTE_FAILURE:
    case actionTypes.REMOVE_USER_ATTRIBUTE_FAILURE:
    case actionTypes.SET_USER_ATTRIBUTE_FAILURE:
    case actionTypes.UPDATE_USER_ATTRIBUTE_FAILURE:
      return (action as ErrorAction).payload.error;
    default:
      return state;
  }
};

const result = (
  state = INITIAL_STATE.result,
  action: AnyAction,
): T.UserAttributesState['result'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ATTRIBUTES_SUCCESS:
      return (action as T.FetchUserAttributesSuccessAction).payload;
    case actionTypes.FETCH_USER_ATTRIBUTE_SUCCESS:
      let currentResult = state;

      if (!currentResult || !Array.isArray(currentResult)) {
        currentResult = [];
      }

      const actionSafe = action as T.FetchUserAttributeSuccessAction;
      const newResult = [...currentResult];

      const index = currentResult.findIndex(
        attribute => attribute?.id === actionSafe.payload.id,
      );

      if (index !== -1) {
        newResult[index] = actionSafe.payload;
      } else {
        newResult.push(actionSafe.payload);
      }

      return newResult;

    default:
      return state;
  }
};

const attributesReducer: Reducer<T.UserAttributesState> = combineReducers({
  isLoading,
  result,
  error,
});

export default attributesReducer;
