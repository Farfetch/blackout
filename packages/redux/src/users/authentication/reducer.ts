import * as actionTypes from './actionTypes.js';
import { type AnyAction, combineReducers, type Reducer } from 'redux';
import reducerFactory from '../../helpers/reducerFactory.js';
import type { AuthenticationState } from './types/index.js';

export const INITIAL_STATE: AuthenticationState = {
  login: {
    error: null,
    isLoading: false,
  },
  logout: {
    error: null,
    isLoading: false,
  },
  register: {
    error: null,
    isLoading: false,
  },
  changePassword: {
    error: null,
    isLoading: false,
  },
  recoverPassword: {
    error: null,
    isLoading: false,
  },
  resetPassword: {
    error: null,
    isLoading: false,
  },
  validateEmail: {
    error: null,
    isLoading: false,
  },
  refreshEmailToken: {
    error: null,
    isLoading: false,
  },
  token: {
    result: null,
    error: null,
    isLoading: false,
  },
};

export const login = reducerFactory('Login', INITIAL_STATE.login, actionTypes);
export const logout = reducerFactory(
  'Logout',
  INITIAL_STATE.logout,
  actionTypes,
);
export const register = reducerFactory(
  'REGISTER',
  INITIAL_STATE.register,
  actionTypes,
);
export const changePassword = reducerFactory(
  'PASSWORD_CHANGE',
  INITIAL_STATE.changePassword,
  actionTypes,
);
export const resetPassword = reducerFactory(
  'PASSWORD_RESET',
  INITIAL_STATE.resetPassword,
  actionTypes,
);
export const recoverPassword = reducerFactory(
  'PASSWORD_RECOVER',
  INITIAL_STATE.recoverPassword,
  actionTypes,
);
export const validateEmail = reducerFactory(
  'VALIDATE_EMAIL',
  INITIAL_STATE.validateEmail,
  actionTypes,
);
export const refreshEmailToken = reducerFactory(
  'REFRESH_EMAIL_TOKEN',
  INITIAL_STATE.refreshEmailToken,
  actionTypes,
);
export const token = (
  state = INITIAL_STATE.token,
  action: AnyAction,
): AuthenticationState['token'] => {
  switch (action?.type) {
    case actionTypes.REMOVE_TOKEN_REQUEST:
    case actionTypes.CREATE_USER_TOKEN_REQUEST:
    case actionTypes.CREATE_CLIENT_CREDENTIALS_TOKEN_REQUEST:
    case actionTypes.REFRESH_TOKEN_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.token.error,
        isLoading: true,
      };
    case actionTypes.REMOVE_TOKEN_SUCCESS:
      return {
        result: INITIAL_STATE.token.result,
        error: INITIAL_STATE.token.error,
        isLoading: false,
      };

    case actionTypes.CREATE_USER_TOKEN_SUCCESS:
    case actionTypes.CREATE_CLIENT_CREDENTIALS_TOKEN_SUCCESS:
    case actionTypes.REFRESH_TOKEN_SUCCESS:
      return {
        result: action?.payload,
        error: INITIAL_STATE.token.error,
        isLoading: false,
      };

    case actionTypes.REMOVE_TOKEN_FAILURE:
    case actionTypes.CREATE_USER_TOKEN_FAILURE:
    case actionTypes.CREATE_CLIENT_CREDENTIALS_TOKEN_FAILURE:
    case actionTypes.REFRESH_TOKEN_FAILURE:
      return {
        ...state,
        error: action?.payload?.error,
        isLoading: false,
      };

    default:
      return state;
  }
};

export const getLogin = (state: AuthenticationState) => state.login;
export const getLogout = (state: AuthenticationState) => state.logout;
export const getRegister = (state: AuthenticationState) => state.register;
export const getChangePassword = (state: AuthenticationState) =>
  state.changePassword;
export const getResetPassword = (state: AuthenticationState) =>
  state.resetPassword;
export const getRecoverPassword = (state: AuthenticationState) =>
  state.recoverPassword;
export const getValidateEmail = (state: AuthenticationState) =>
  state.validateEmail;
export const getRefreshEmailToken = (state: AuthenticationState) =>
  state.refreshEmailToken;
export const getToken = (state: AuthenticationState) => state.token;

/**
 * Reducer for authentication state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const authenticationReducer: Reducer<AuthenticationState> = combineReducers({
  login,
  logout,
  register,
  changePassword,
  resetPassword,
  recoverPassword,
  validateEmail,
  refreshEmailToken,
  token,
});

export default authenticationReducer;
