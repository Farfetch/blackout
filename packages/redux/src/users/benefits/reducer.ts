import * as actionTypes from './actionTypes';
import reducerFactory from '../../helpers/reducerFactory';
import type { AnyAction } from 'redux';
import type { StoreState } from '../../types/storeState.types';
import type { UserEntity } from '../../entities';

export const INITIAL_STATE = {
  error: null,
  isLoading: false,
};

export const entitiesMapper = {
  [actionTypes.FETCH_USER_BENEFITS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { user: currentUser } = state;

    if (!currentUser) {
      return state;
    }

    const { benefits } = action.payload.entities;
    // Add benefits reference to user entity
    const user = {
      ...(state.user as NonNullable<UserEntity>),
      benefits: action.payload.result,
    };

    return {
      ...state,
      user,
      benefits,
    };
  },
};

export default reducerFactory(
  'FETCH_USER_BENEFITS',
  INITIAL_STATE,
  actionTypes,
);
