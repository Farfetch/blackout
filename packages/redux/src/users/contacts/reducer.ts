import * as actionTypes from './actionTypes.js';
import reducerFactory from '../../helpers/reducerFactory.js';
import type { AnyAction } from 'redux';
import type { StoreState } from '../../types/storeState.types.js';
import type { UserEntity } from '../../entities/index.js';

export const INITIAL_STATE = {
  error: null,
  isLoading: false,
};

export const entitiesMapper = {
  [actionTypes.FETCH_USER_CONTACTS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { user: currentUser } = state;

    if (!currentUser) {
      return state;
    }

    const { contacts } = action.payload.entities;
    // Add contacts reference to user entity
    const user = {
      ...(state.user as NonNullable<UserEntity>),
      contacts: action.payload.result,
    };

    return {
      ...state,
      user,
      contacts,
    };
  },
};

export default reducerFactory(
  ['FETCH_USER_CONTACTS', 'FETCH_USER_CONTACT'],
  INITIAL_STATE,
  actionTypes,
);
