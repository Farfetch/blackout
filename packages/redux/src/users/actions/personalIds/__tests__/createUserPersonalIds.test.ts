import { actionTypes } from '../../../';
import { createUserPersonalIds } from '../';
import { INITIAL_STATE } from '../../../reducer';
import { mockPostPersonalIdsResponse } from 'tests/__fixtures__/users';
import { mockStore } from '../../../../../tests';
import { postUserPersonalIds } from '@farfetch/blackout-client';
import find from 'lodash/find';

jest.mock('@farfetch/blackout-client', () => ({
  ...jest.requireActual('@farfetch/blackout-client'),
  postUserPersonalIds: jest.fn(),
}));

const usersMockStore = (state = {}) =>
  mockStore({ users: INITIAL_STATE }, state);

describe('createUserPersonalIds action creator', () => {
  let store = usersMockStore();
  const data = {
    backImageId: '',
    frontImageId: '',
    idNumber: '',
    name: '',
  };
  const userId = 12345;
  const config = {
    'X-SUMMER-RequestId': 'test',
  };
  const expectedConfig = {
    'X-SUMMER-RequestId': 'test',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    store = usersMockStore();
  });

  it('should create the correct actions for when the create personal ids procedure fails', async () => {
    const expectedError = new Error('create user attributes error');

    (postUserPersonalIds as jest.Mock).mockRejectedValueOnce(expectedError);
    expect.assertions(4);

    try {
      await store.dispatch(createUserPersonalIds(userId, data, config));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(postUserPersonalIds).toHaveBeenCalledTimes(1);
      expect(postUserPersonalIds).toHaveBeenCalledWith(
        userId,
        data,
        expectedConfig,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          { type: actionTypes.CREATE_USER_PERSONAL_IDS_REQUEST },
          {
            type: actionTypes.CREATE_USER_PERSONAL_IDS_FAILURE,
            payload: { error: expectedError },
          },
        ]),
      );
    }
  });

  it('should create the correct actions for when the create personal ids procedure is successful', async () => {
    (postUserPersonalIds as jest.Mock).mockResolvedValueOnce(
      mockPostPersonalIdsResponse,
    );
    await store.dispatch(createUserPersonalIds(userId, data, config));

    const actionResults = store.getActions();

    expect(postUserPersonalIds).toHaveBeenCalledTimes(1);
    expect(postUserPersonalIds).toHaveBeenCalledWith(
      userId,
      data,
      expectedConfig,
    );

    expect(actionResults).toMatchObject([
      { type: actionTypes.CREATE_USER_PERSONAL_IDS_REQUEST },
      {
        payload: mockPostPersonalIdsResponse,
        type: actionTypes.CREATE_USER_PERSONAL_IDS_SUCCESS,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.CREATE_USER_PERSONAL_IDS_SUCCESS,
      }),
    ).toMatchSnapshot('create personal ids success payload');
  });
});
