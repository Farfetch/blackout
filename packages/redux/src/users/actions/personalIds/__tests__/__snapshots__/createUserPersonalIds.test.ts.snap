// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUserPersonalIds action creator should create the correct actions for when the create personal ids procedure is successful: create personal ids success payload 1`] = `
Object {
  "payload": Object {
    "backImageId": "string",
    "expiryDate": "string",
    "frontImageId": "string",
    "id": "string",
    "idNumber": "string",
    "name": "string",
  },
  "type": "@farfetch/blackout-client/CREATE_USER_PERSONAL_IDS_SUCCESS",
}
`;
